src/mindlessgen/molecules/generate_molecule.py

```diff
@@ -426,3 +456,106 @@ def check_distances(
+    num_atoms: int,
+    nel: int,
+    valid_elems: set[int],
+    verbosity: int,
+) -> np.ndarray:
+    """
+    Correct the number of electrons if a fixed charge is given and the molecule is not closed shell.
+    """
+    f_elem, ligand_protons = calculate_ligand_protons(natoms, nel)
+    # If f block elements are included, cerrect only if the remaning ligand protons are uneven
+    if f_elem and ligand_protons % 2 != 0:
+        if 0 in valid_elems:
+            natoms = fixed_charge_hydrogen_correction(
+                cfg, natoms, num_atoms, valid_elems, verbosity
+            )
+            return natoms
+        else:
+            natoms = fixed_charge_elem_correction(
+                cfg, natoms, num_atoms, valid_elems, verbosity
+            )
+            return natoms
+    # If f block elements are not included, correct if the number of electrons is uneven
+    elif nel % 2 != 0 and f_elem is False:
+        if 0 in valid_elems:
+            natoms = fixed_charge_hydrogen_correction(
+                cfg, natoms, num_atoms, valid_elems, verbosity
+            )
+            return natoms
+        else:
+            natoms = fixed_charge_elem_correction(
+                cfg, natoms, num_atoms, valid_elems, verbosity
+            )
+            return natoms
+    return natoms
+
+
+def fixed_charge_hydrogen_correction(
+    cfg: GenerateConfig,
+    natoms: np.ndarray,
+    num_atoms: int,
+    valid_elems: set[int],
+    verbosity: int,
+) -> np.ndarray:
+    """
+    Correct the number of hydrogen atoms if the number of electrons is odd.
+    """
+    min_count, max_count = cfg.element_composition.get(0, (0, cfg.max_num_atoms))
+    if max_count is None:
+        max_count = cfg.max_num_atoms
+    # Check if adding or removing hydrogen atoms is possible
+    if natoms[0] < max_count and num_atoms < cfg.max_num_atoms:
+        natoms[0] += 1
+        if verbosity > 1:
+            print("Adding hydrogen atom...")
+        return natoms
+    elif natoms[0] > min_count and num_atoms > cfg.min_num_atoms:
+        natoms[0] -= 1
+        if verbosity > 1:
+            print("Removing hydrogen atom...")
+        return natoms
+    else:
+        natoms = fixed_charge_elem_correction(
+            cfg, natoms, num_atoms, valid_elems, verbosity
+        )
+        return natoms
+
+
+def fixed_charge_elem_correction(
+    cfg: GenerateConfig,
+    natoms: np.ndarray,
+    num_atoms: int,
+    valid_elems: set[int],
+    verbosity: int,
+) -> np.ndarray:
+    """
+    Correct the number of atoms if the number of electrons is odd and hydrogen can not be added.
+    """
+    # All other elements
+    odd_atoms = [elem for elem in valid_elems if elem % 2 == 0 and elem != 0]
+    for random_elem in np.random.permutation(odd_atoms):
+        min_count, max_count = cfg.element_composition.get(
+            random_elem, (0, cfg.max_num_atoms)
+        )
+        if max_count is None:
+            max_count = cfg.max_num_atoms
```

Comment:
If an asterix is given within the element composition this seems to not work with `dict.get()`

Reply:
But why? Due to the following line, an asterisk is converted to `None` anyway.
```python
            min_count = None if min_count == "*" else int(min_count)
            max_count = None if max_count == "*" else int(max_count)
```

Reply:
Yes This value is passed as None and this is then the `max_count` but this would not add anything. The `dict.get()` function just set `max_count = cfg.max_num_atoms` if nothing would be given as `cfg.max_num_atoms`.